File: src_py/crawlers/mss.py
"""중소벤처기업부 (mss.go.kr) 보도자료 크롤러."""

from __future__ import annotations

import logging
import re
from datetime import datetime

from bs4 import BeautifulSoup

from .base import Article, BaseCrawler

logger = logging.getLogger(__name__)

BASE_URL = "https://www.mss.go.kr"
# cbIdx=86: 보도자료 게시판
LIST_URL = f"{BASE_URL}/site/smba/ex/bbs/List.do"
BOARD_ID = "86"


class MSSCrawler(BaseCrawler):
    """중소벤처기업부 보도자료 크롤러."""

    name = "중소벤처기업부"

    def crawl(self) -> list[Article]:
        params = {"cbIdx": BOARD_ID, "pageIndex": "1"}
        resp = self.fetch(LIST_URL, params=params)
        if resp is None:
            return []

        resp.encoding = "utf-8"
        soup = BeautifulSoup(resp.text, "html.parser")
        articles: list[Article] = []

        rows = self._find_rows(soup)
        for row in rows:
            article = self._parse_row(row)
            if article:
                articles.append(article)

        logger.info("[%s] %d건 수집 완료", self.name, len(articles))
        return articles

    def _find_rows(self, soup: BeautifulSoup) -> list:
        selectors = [
            "table.boardList tbody tr",
            "table.bbs_list tbody tr",
            "div.board_list table tbody tr",
            "table.tbl_type tbody tr",
            "table tbody tr",
        ]
        for sel in selectors:
            rows = soup.select(sel)
            if rows:
                return rows
        return []

    def _parse_row(self, row) -> Article | None:
        cols = row.select("td")
        if len(cols) < 3:
            return None

        link = row.select_one("a")
        if not link:
            return None

        title = link.get_text(strip=True)
        if not title:
            return None

        url = self._extract_url(link)
        date = self._extract_date(cols)

        return Article(title=title, url=url, source=self.name, date=date)

    def _extract_url(self, link) -> str:
        href = link.get("href", "")
        if href and href.startswith("http"):
            return href
        if href and href.startswith("/"):
            return BASE_URL + href

        # onclick 핸들러 파싱
        onclick = link.get("onclick", "")
        if onclick:
            # fn_detail('86', '12345') 패턴
            match = re.search(r"'(\d+)'\s*,\s*'(\d+)'", onclick)
            if match:
                board_id, article_id = match.groups()
                return f"{BASE_URL}/site/smba/ex/bbs/View.do?cbIdx={board_id}&bcIdx={article_id}"
            # 단순 숫자 ID
            match = re.search(r"(\d{4,})", onclick)
            if match:
                return f"{BASE_URL}/site/smba/ex/bbs/View.do?cbIdx={BOARD_ID}&bcIdx={match.group(1)}"

        return LIST_URL

    def _extract_date(self, cols: list) -> str:
        """컬럼들에서 날짜 형식을 찾아 반환."""
        for col in cols:
            text = col.get_text(strip=True)
            match = re.match(r"^(\d{4})[.\-](\d{1,2})[.\-](\d{1,2})$", text)
            if match:
                y, m, d = match.groups()
                try:
                    dt = datetime(int(y), int(m), int(d))
                    return dt.strftime("%Y-%m-%d")
                except ValueError:
                    pass
        return datetime.now().strftime("%Y-%m-%d")

File: src/lib/crawlers/mss.ts
import { extractDate } from '@/lib/utils';
import { BaseCrawler } from './base';
import type { CheerioAPI, AnyNode, Cheerio } from './base';
import type { RawArticle } from './types';

const BASE_URL = 'https://www.mss.go.kr';
const LIST_URL = `${BASE_URL}/site/smba/ex/bbs/List.do`;
const BOARD_ID = '86';

export class MSSCrawler extends BaseCrawler {
  name = '중소벤처기업부';

  async crawl(): Promise<RawArticle[]> {
    const html = await this.fetch(`${LIST_URL}?cbIdx=${BOARD_ID}&pageIndex=1`);
    if (!html) return [];

    const $ = this.parse(html);
    const articles: RawArticle[] = [];

    const rows = this.findRows($);
    for (const row of rows) {
      const article = this.parseRow($, row);
      if (article) articles.push(article);
    }

    console.log(`[${this.name}] ${articles.length}건 수집 완료`);
    return articles;
  }

  private findRows($: CheerioAPI): AnyNode[] {
    const selectors = [
      'table.boardList tbody tr',
      'table.bbs_list tbody tr',
      'div.board_list table tbody tr',
      'table.tbl_type tbody tr',
      'table tbody tr',
    ];
    for (const sel of selectors) {
      const rows = $(sel).toArray();
      if (rows.length > 0) return rows;
    }
    return [];
  }

  private parseRow($: CheerioAPI, row: AnyNode): RawArticle | null {
    const $row = $(row);
    const cols = $row.find('td');
    if (cols.length < 3) return null;

    const link = $row.find('a').first();
    if (!link.length) return null;

    const title = link.text().trim();
    if (!title) return null;

    const url = this.extractUrl(link);
    const date = this.extractDateFromCols($, cols);

    return { title, url, source: this.name, date, deadline: '' };
  }

  private extractUrl(link: Cheerio<AnyNode>): string {
    const href = link.attr('href') ?? '';
    if (href.startsWith('http')) return href;
    if (href.startsWith('/')) return BASE_URL + href;

    const onclick = link.attr('onclick') ?? '';
    if (onclick) {
      const match = onclick.match(/'(\d+)'\s*,\s*'(\d+)'/);
      if (match) {
        return `${BASE_URL}/site/smba/ex/bbs/View.do?cbIdx=${match[1]}&bcIdx=${match[2]}`;
      }
      const idMatch = onclick.match(/(\d{4,})/);
      if (idMatch) {
        return `${BASE_URL}/site/smba/ex/bbs/View.do?cbIdx=${BOARD_ID}&bcIdx=${idMatch[1]}`;
      }
    }

    return LIST_URL;
  }

  private extractDateFromCols($: CheerioAPI, cols: Cheerio<AnyNode>): string {
    for (let i = 0; i < cols.length; i++) {
      const text = $(cols[i]).text().trim();
      const match = text.match(/^(\d{4})[.\-](\d{1,2})[.\-](\d{1,2})$/);
      if (match) {
        const [, y, m, d] = match;
        try {
          const dt = new Date(Number(y), Number(m) - 1, Number(d));
          if (!isNaN(dt.getTime())) return dt.toISOString().slice(0, 10);
        } catch { /* skip */ }
      }
    }
    return extractDate('');
  }
}
